### PR TITLE
gateway: skip un-managed types in deploymentcontroller

### PR DESCRIPTION
**Pull request by howardjohn**
fixes https://github.com/istio/istio/issues/43439

This adds a check to apply that the resource either does not exist or has the managed label. This ensures that we do not overwrite users existing resources, which could cause an outage, etc.

A few concerns though...:
* Its racy; if a user deploys a folder of YAML with a Gateway and a Service, depending on the ordering we may manage it or may not. With Apply semantics, their `apply service.yaml` will NOT make it unmanaged since it would merge labels with the controller
* Its also racy even if user is careful due to implementation. We are using Apply, which cannot do preconditions. So we have a TOCTOU race where we may see it is "managed" (or not existing) and decide to apply, but that changes before we actually Apply. AFAIK k8s can't do precondition checks on this type of operation?

### CODE REVIEW

**Comment by therealmitchconnors** [comment]
Generally speaking, if a controller wants a resource to exist, it is expected to overwrite any pre existing resource with that name.  That is why controllers should always specify force=true on their Apply() functions.  We can add a pre-upgrade check to see if any of the user's existing deployments could potentially collide with waypoint names...

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 1]
```diff
@@ -431,6 +432,30 @@ func (d *DeploymentController) apply(controller string, yml string) error {
...
+		// protect against overwrites though.
+		return true
+	}
+	obj, f, _ := store.GetByKey(controllers.KeyFunc(name, namespace))
+	if f {
+		m, ok := obj.(controllers.Object)
+		if ok {
+			_, managed := m.GetLabels()[constants.ManagedGatewayLabel]
+			// If object already exists, we can only manage it if it has the label
+			return managed
```
Comment (costinm) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:452]: Do we need to worry about revision/labels ? Or is this protected by a master election - and winner manages everything ?

**Review by costinm — APPROVED** [review]
Looks good.

**Comment by howardjohn** [comment]
> Generally speaking, if a controller wants a resource to exist, it is expected to overwrite any pre existing resource with that name. That is why controllers should always specify force=true on their Apply() functions. We can add a pre-upgrade check to see if any of the user's existing deployments could potentially collide with waypoint names...

This is not a one-time issue on upgrade. This impacts both waypoints and ingress types, and can happen at anytime beyond upgrade. Both of these can also be named arbitrarily. The intent of this PR is to avoid disaster scenario where we overwrite their prod service and cause an outage. I think other tools, like helm, do similar (they won't take over a resource without the helm ownership labels).

My main concern personally was the api-machinery. I think I have convinced myself this is safe after some tweaks + discussion with k8s folks (see linked issue, in the code). So I think that leaves this discussion of whether it is _behaviorally_ correct? WDYT

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 2]
```diff
@@ -419,6 +413,12 @@ func (d *DeploymentController) apply(controller string, yml string) error {
 	if err != nil {
 		return err
 	}
+	canManage, resourceVersion := d.canManage(gvr, us.GetName(), us.GetNamespace())
+	if !canManage {
+		log.Debugf("skipping %v/%v/%v, already managed", gvr, us.GetName(), us.GetNamespace())
+		return nil
```
Comment (therealmitchconnors) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:419]: In this case, we should update the status to reflect that the gateway has not been implemented.  That might be happing elsewhere, but I can't tell from reading the code...

**Review by therealmitchconnors — COMMENTED** [review]
Recommend adding a test, either unit or e2e.

**Pull request merged**